### PR TITLE
MM-33359 corrected unread replies update

### PR DESCRIPTION
**Pull request by reflog**
#### Summary

When user replies to a thread, the thread's 'read' state for that specific user should be updated, otherwise unreadReplies will include their 'self-reply'.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-33359
#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: store/sqlstore/thread_store.go [hunk 1]
```diff
@@ -501,14 +501,17 @@ func (s *SqlThreadStore) DeleteMembershipForUser(userId string, postId string) e
 	return nil
 }
 
-func (s *SqlThreadStore) CreateMembershipIfNeeded(userId, postId string, following, incrementMentions, updateFollowing bool) error {
+func (s *SqlThreadStore) MaintainMembership(userId, postId string, following, incrementMentions, updateFollowing, updateViewedTimestamp bool) error {
 	membership, err := s.GetMembershipForUser(userId, postId)
 	now := utils.MillisFromTime(time.Now())
 	if err == nil {
-		if (updateFollowing && !membership.Following || membership.Following != following) || incrementMentions {
+		if (updateFollowing && !membership.Following || membership.Following != following) || incrementMentions || updateViewedTimestamp {
```
Comment (streamer45) [inline store/sqlstore/thread_store.go:508]: This has gotten a bit hardcore. Can we add a comment explaining the logic?

File: store/sqlstore/thread_store.go [hunk 2]
```diff
@@ -501,14 +501,17 @@ func (s *SqlThreadStore) DeleteMembershipForUser(userId string, postId string) e
 	return nil
 }
 
-func (s *SqlThreadStore) CreateMembershipIfNeeded(userId, postId string, following, incrementMentions, updateFollowing bool) error {
+func (s *SqlThreadStore) MaintainMembership(userId, postId string, following, incrementMentions, updateFollowing, updateViewedTimestamp bool) error {
```
Comment (agnivade) [inline store/sqlstore/thread_store.go:504]: Can we add a test please? :)

File: app/notification.go [hunk 3]
```diff
@@ -183,14 +183,8 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
...
-				incrementMentions := false
-				for mid := range mentions.Mentions {
-					if userID == mid {
-						incrementMentions = true
-						break
-					}
-				}
-				nErr := a.Srv().Store.Thread().CreateMembershipIfNeeded(userID, post.RootId, true, incrementMentions, *a.Config().ServiceSettings.ThreadAutoFollow)
+				_, incrementMentions := mentions.Mentions[userID]
+				nErr := a.Srv().Store.Thread().MaintainMembership(userID, post.RootId, true, incrementMentions, *a.Config().ServiceSettings.ThreadAutoFollow, userID == post.UserId)
```
Comment (agnivade) [inline app/notification.go:187]: Let's take this opportunity and change `nErr` to `err`

**Review by agnivade — CHANGES_REQUESTED** [review]

File: store/sqlstore/thread_store.go (hunk 2 above)
Comment (reflog) [reply store/sqlstore/thread_store.go:504]: test cases already exist in store/storetest/thread_store.go . I've just renamed the function to be clearer and added a param to update a timestamp

File: store/sqlstore/thread_store.go (hunk 1 above)
Comment (reflog) [reply store/sqlstore/thread_store.go:508]: done

File: store/sqlstore/thread_store.go (hunk 2 above)
Comment (reflog) [reply store/sqlstore/thread_store.go:504]: @agnivade added

**Review by agnivade — APPROVED** [review]
Looks great

File: store/sqlstore/thread_store.go [hunk 4]
```diff
@@ -501,14 +501,21 @@ func (s *SqlThreadStore) DeleteMembershipForUser(userId string, postId string) e
...
+	// c. user viewed a thread
 	if err == nil {
-		if (updateFollowing && !membership.Following || membership.Following != following) || incrementMentions {
+		if (updateFollowing && !membership.Following || membership.Following != following) || incrementMentions || updateViewedTimestamp {
 			if updateFollowing {
 				membership.Following = following
 			}
+			if updateViewedTimestamp {
+				membership.LastViewed = now
+			}
```
Comment (streamer45) [inline store/sqlstore/thread_store.go:518]: So why are we checking for `membership.Following != following` but then rely solely on `updateFollowing` to do the update? Do we have redundant logic/variables or are we possibly missing a case?

**Pull request merged**